Neomake not working after update (windows 10, vim8)

Since I updated neomake yesterday my linters do not work anymore.
`:PlugDiff` says the last update looked like this:
```
* a23ce39 Travis: enable/enforce vimlint, drop vimlint-errors (#987) (3 days ago)
[... many lines in between]
* 860b3e8 Add swiftc maker for swift filetype (8 weeks ago)
```

Logfile and verbose output (also can be seen below) looks like this:
```
Neomake [4536.698]: Running makers: eslint
Neomake [4536.710]: [30] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4536.719]: [30.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4536.721]: [30.0] Vim channel: 0
```
The followings are working just fine:
```
:!eslint %
:!lessc %
```

### Expected behavior

My linters (eslint, custom lessc) respond with errors that are then marked in vim.

### Steps to reproduce

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Nov  6 2016 04:20:56)
"\Users\foellerich\Git\<path to>\<file>.less" [unix] 465L, 10124C
"cart.less" [unix] 465L, 10128C written
"cart.js" 236 lines --18%--
Neomake [4092.203]: Running makers: eslint
Neomake [4092.213]: [26] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4092.218]: [26.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4092.219]: [26.0] Vim channel: 0
Neomake [4114.902]: Running makers: eslint
Neomake [4114.903]: [27] Restarting already running job (26.0) for the same maker.
Neomake [4114.904]: [26.0] Stopping job
Neomake [4114.905]: [26.0] job_stop: job was not running anymore
Neomake [4118.695]: Running makers: eslint
Neomake [4118.706]: [28] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4118.711]: [28.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4118.712]: [28.0] Vim channel: 0
"cart.less" 465 lines --0%--
Neomake [4152.019]: Running makers: lessc
Neomake [4152.026]: [29] Starting async job: ['lessc', '--no-color', '--lint', '--silent', 'C:\Users\foellerich\Git\<path to>\<file>.less']
Neomake [4152.034]: [29.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4152.036]: [29.0] Vim channel: 0
"cart.js" 236 lines --19%--
"cart.js" [unix] 236L, 9155C written
Neomake [4536.698]: Running makers: eslint
Neomake [4536.710]: [30] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4536.719]: [30.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4536.721]: [30.0] Vim channel: 0
```

### Output of the ":verb NeomakeInfo" command


#### Neomake debug information
Async support: 1
Current filetype: javascript


##### Enabled makers
For the current filetype (with :Neomake):
 - eslint
   errorformat: '%E%f: line %l\, col %c\, Error - %m,%W%f: line %l\, col %c\, Warning - %m'
   args: ['-f', 'compact']
   get_argv: function('57', {'get_argv': function('57')})
   exe: 'eslint'
NOTE: you can define g:neomake_javascript_enabled_makers to configure it (or b:neomake_javascript_enabled_makers).


For the project (with :Neomake!): None.
NOTE: you can define g:neomake_enabled_makers to configure it.


Default maker settings:
 - remove_invalid_entries: 0
 - buffer_output: 1


##### Settings
```
g:neomake_open_list = 2
g:neomake_less_lessc_maker = {'errorformat': '%m in %f on line %l\, column %c:,%m in %f:%l:%c,%-G%.%#', 'args': ['--no-color', '--lint', '--silent']}
g:neomake_javascript_enabled_makers = ['eslint']
g:neomake_place_signs = 1
g:neomake_logfile = 'C:\tmp\neomake.log'
g:neomake_less_enabled_makers = ['lessc']
g:neomake_verbose = 3


shell: C:\WINDOWS\system32\cmd.exe
shellcmdflag: /c
Windows: 1
```


#### :version
```

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Nov  6 2016 04:20:56)
MS-Windows 64-bit GUI version with OLE support
Included patches: 1-66
Compiled by hp@HP-PC
Huge version with GUI.  Features included (+) or not (-):
+acl                +clientserver       +cursorbind         +eval               +gettext/dyn        +libcall            +mouse              +path_extra         +reltime            +syntax             +timers             +vreplace           
+arabic             +clipboard          +cursorshape        +ex_extra           -hangul_input       +linebreak          +mouseshape         +perl/dyn           +rightleft          +tag_binary         +title              +wildignore         
+autocmd            +cmdline_compl      +dialog_con_gui     +extra_search       +iconv/dyn          +lispindent         +multi_byte_ime/dyn +persistent_undo    +ruby/dyn           +tag_old_static     +toolbar            +wildmenu           
+balloon_eval       +cmdline_hist       +diff               +farsi              +insert_expand      +listcmds           +multi_lang         -postscript         +scrollbind         -tag_any_white      +user_commands      +windows            
+browse             +cmdline_info       +digraphs           +file_in_path       +job                +localmap           +mzscheme/dyn       +printer            +signs              +tcl/dyn            +vertsplit          +writebackup        
++builtin_terms     +comments           +directx            +find_in_path       +jumplist           +lua/dyn            +netbeans_intg      +profile            +smartindent        -termguicolors      +virtualedit        -xfontset           
+byte_offset        +conceal            -dnd                +float              +keymap             +menu               +num64              +python/dyn         +startuptime        -tgetent            +visual             -xim                
+channel            +cryptv             -ebcdic             +folding            +lambda             +mksession          +ole                +python3/dyn        +statusline         -termresponse       +visualextra        +xpm_w32            
+cindent            +cscope             +emacs_tags         -footer             +langmap            +modify_fname       +packages           +quickfix           -sun_workshop       +textobjects        +viminfo            -xterm_save         
   system vimrc file: "$VIM\vimrc"
     user vimrc file: "$HOME\_vimrc"
 2nd user vimrc file: "$HOME\vimfiles\vimrc"
 3rd user vimrc file: "$VIM\_vimrc"
      user exrc file: "$HOME\_exrc"
  2nd user exrc file: "$VIM\_exrc"
  system gvimrc file: "$VIM\gvimrc"
    user gvimrc file: "$HOME\_gvimrc"
2nd user gvimrc file: "$HOME\vimfiles\gvimrc"
3rd user gvimrc file: "$VIM\_gvimrc"
       defaults file: "$VIMRUNTIME\defaults.vim"
    system menu file: "$VIMRUNTIME\menu.vim"
Compilation: cl -c /W3 /nologo  -I. -Iproto -DHAVE_PATHDEF -DWIN32  -DFEAT_CSCOPE -DFEAT_NETBEANS_INTG -DFEAT_JOB_CHANNEL   -DFEAT_XPM_W32   -DWINVER=0x0501 -D_WIN32_WINNT=0x0501  /Fo.\ObjGXOULYHTRZAMD64/ -DHAVE_STDINT_H /Ox /GL -DNDEBUG  /Zl /MT -DFEAT_OLE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DGLOBAL_IME -DFEAT_MBYTE -DFEAT_GUI_W32 -DFEAT_DIRECTX -DDYNAMIC_DIRECTX -DDYNAMIC_ICONV -DDYNAMIC_GETTEXT -DFEAT_TCL -DDYNAMIC_TCL -DDYNAMIC_TCL_DLL=\"tcl86.dll\" -DDYNAMIC_TCL_VER=\"8.6\" -DFEAT_LUA -DDYNAMIC_LUA -DDYNAMIC_LUA_DLL=\"lua53.dll\" -DFEAT_PYTHON -DDYNAMIC_PYTHON -DDYNAMIC_PYTHON_DLL=\"python27.dll\" -DFEAT_PYTHON3 -DDYNAMIC_PYTHON3 -DDYNAMIC_PYTHON3_DLL=\"python35.dll\" -DFEAT_MZSCHEME -I "C:\Users\hp\Desktop\racket-x64\include" -DDYNAMIC_MZSCHEME -DDYNAMIC_MZSCH_DLL=\"libracketxxxxxxx.dll\" -DDYNAMIC_MZGC_DLL=\"libmzgcxxxxxxx.dll\" -DFEAT_PERL -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DDYNAMIC_PERL -DDYNAMIC_PERL_DLL=\"perl524.dll\" -DFEAT_RUBY -DDYNAMIC_RUBY -DDYNAMIC_RUBY_VER=23 -DDYNAMIC_RUBY_DLL=\"x64-msvcrt-ruby230.dll\" -DFEAT_HUGE /Fd.\ObjGXOULYHTRZAMD64/ /Zi
Linking: link /RELEASE /nologo /subsystem:windows /LTCG:STATUS oldnames.lib kernel32.lib advapi32.lib shell32.lib gdi32.lib  comdlg32.lib ole32.lib uuid.lib /machine:AMD64 gdi32.lib version.lib   winspool.lib comctl32.lib advapi32.lib shell32.lib  /machine:AMD64 /nodefaultlib libcmt.lib oleaut32.lib user32.lib  /nodefaultlib:lua53.lib  /STACK:8388608  /nodefaultlib:python27.lib /nodefaultlib:python35.lib   "C:\TCL\lib\tclstub86.lib" WSock32.lib .\xpm\x64\lib\libXpm.lib /PDB:gvim.pdb -debug
```


#### :messages
```
Messages maintainer: Bram Moolenaar <Bram@vim.org>
"~\Git\<path to>\<file>.js" [unix] 426L, 17479C
"cart-test.js" [unix] 426L, 17479C written
0

"cart-test.js" [unix] 427L, 17480C written
0

"cart-test.js" [unix] 429L, 17482C written

"help.txt" [RO][unix] 225L, 8437C
"~\Git\<path to>\<file>.js" [unix] 236L, 9156C
"cart.js" [unix] 236L, 9155C written
1 change; before #1  7 seconds ago
"cart.js" [unix] 236L, 9156C written
"cart-test.js" 429 lines --20%--
"cart-test.js" [unix] 426L, 17479C written
0

0

"cart-test.js" [unix] 426L, 17481C written
"cart.js" 236 lines --16%--
E516: No buffers were deleted
"cart.js" [unix] 236L, 9155C written
E492: Not an editor command: neomake

"~\Git\<path to>\<file>.less" [unix] 53L, 1230C
"alerts.less" [unix] 53L, 1229C written
1 change; before #1  10 seconds ago
"alerts.less" [unix] 53L, 1230C written
"~\Git\<path to>\<file>.less" [unix] 465L, 10124C
"cart.less" [unix] 465L, 10128C written
"cart.less" [unix] 465L, 10124C written
"alerts.less" 53 lines --9%--
"cart.js" 236 lines --18%--
E492: Not an editor command: neomake eslint

E492: Not an editor command: neomake
E492: Not an editor command: neomake
0 neomake_12 eslint
0 neomake_12 eslint
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Nov  6 2016 04:20:56)
"\Users\foellerich\Git\<path to>\<file>.less" [unix] 465L, 10124C
"cart.less" [unix] 465L, 10128C written
"cart.js" 236 lines --18%--
Neomake [4092.203]: Running makers: eslint
Neomake [4092.213]: [26] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4092.218]: [26.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4092.219]: [26.0] Vim channel: 0
Neomake [4114.902]: Running makers: eslint
Neomake [4114.903]: [27] Restarting already running job (26.0) for the same maker.
Neomake [4114.904]: [26.0] Stopping job
Neomake [4114.905]: [26.0] job_stop: job was not running anymore
Neomake [4118.695]: Running makers: eslint
Neomake [4118.706]: [28] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4118.711]: [28.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4118.712]: [28.0] Vim channel: 0
"cart.less" 465 lines --0%--
Neomake [4152.019]: Running makers: lessc
Neomake [4152.026]: [29] Starting async job: ['lessc', '--no-color', '--lint', '--silent', 'C:\Users\foellerich\Git\<path to>\<file>.less']
Neomake [4152.034]: [29.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4152.036]: [29.0] Vim channel: 0
"cart.js" 236 lines --19%--
"cart.js" [unix] 236L, 9155C written
Neomake [4536.698]: Running makers: eslint
Neomake [4536.710]: [30] Starting async job: ['eslint', '-f', 'compact', 'C:\Users\foellerich\Git\<path to>\<file>.js']
Neomake [4536.719]: [30.0] Vim job: {'status': 'fail', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel fail, 'process': 0}
Neomake [4536.721]: [30.0] Vim channel: 0
```

set shellslash gives proper slashes when doing ES6 imports or using any other filepath completion on windows. 

This is the output with noshellslash and the patch.

```
Neomake [     ]: [2] Running makers: tsc (auto), tslint (auto)
Neomake [     ]: "tsc" "--project" "c:\path\to\tsconfig.json" "--noEmit"
Neomake [     ]: C:\WINDOWS\system32\cmd.exe /c """tsc"" ""--project"" ""c:\path\to\tsconfig.json"" ""--noEmit"""
Neomake [     ]: [2] Starting async job: 'C:\WINDOWS\system32\cmd.exe /c """tsc"" ""--project"" ""c:\path\to\tsconfig.json"" ""--noEmit"""'
Neomake [     ]: [2.2] Vim job: {'status': 'run', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel 2 open, 'process': 16860}
Neomake [     ]: [2.2] Vim channel: {'status': 'open', 'id': 2, 'in_io': 'pipe', 'err_mode': 'RAW', 'in_status': 'open', 'out_io': 'pipe', 'err_io': 'pipe', 'err_status': 'open', 'out_timeout': 2000, 'out_status': 'open', 'out_mode': 'RAW', 'in_mode': 'RAW', 'in_timeout': 2000, 'err_timeout': 2000}
Neomake [+0.01]: "tslint" "c:\path\to\broen-file.ts" "--format verbose" "c:\path\to\broen-file.ts"
Neomake [     ]: C:\WINDOWS\system32\cmd.exe /c """tslint"" ""c:\path\to\broen-file.ts"" ""--format verbose"" ""c:\path\to\broen-file.ts"""
Neomake [     ]: [2] Starting async job: 'C:\WINDOWS\system32\cmd.exe /c """tslint"" ""c:\path\to\broen-file.ts"" ""--format verbose"" ""c:\path\to\broen-file.ts"""'
Neomake [     ]: [2.3] Vim job: {'status': 'run', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel 3 open, 'process': 1260}
Neomake [     ]: [2.3] Vim channel: {'status': 'open', 'id': 3, 'in_io': 'pipe', 'err_mode': 'RAW', 'in_status': 'open', 'out_io': 'pipe', 'err_io': 'pipe', 'err_status': 'open', 'out_timeout': 2000, 'out_status': 'open', 'out_mode': 'RAW', 'in_mode': 'RAW', 'in_timeout': 2000, 'err_timeout': 2000}
Neomake [+0.49]: [2.3] stderr: tslint: ['module.js:474', '    throw err;', '    ^', '', 'Error: Cannot find module ''c:\path\where\node_modules\doesnt\exist\node_modules\tslint\bin\tslint''', '    at Function.Module._resolveFilename (module.js:472:15)', '    at Function.Module._load (module.js:420:25)', '    at Module.runMain (module.js:607:10)', '    at run (bootstrap_node.js:382:7)', '    at startup (bootstrap_node.js:137:9)', '    at bootstrap_node.js:497:3', '']
Neomake [     ]: [2.2] stderr: tsc: ['module.js:474', '    throw err;', '    ^', '', 'Error: Cannot find module ''c:\path\where\node_modules\doesnt\exist\node_modules\typescript\bin\tsc''', '    at Function.Module._resolveFilename (module.js:472:15)', '    at Function.Module._load (module.js:420:25)', '    at Module.runMain (module.js:607:10)', '    at run (bootstrap_node.js:382:7)', '    at startup (bootstrap_node.js:137:9)', '    at bootstrap_node.js:497:3', '']
Neomake [+0.10]: [2.3] exit: tslint: 1
Neomake [     ]: [2.3] tslint: completed with exit code 1.
Neomake [     ]: [2.3] Output left to be processed, not cleaning job yet.
Neomake [     ]: [2.2] exit: tsc: 1
Neomake [     ]: [2.2] tsc: completed with exit code 1.
Neomake [     ]: [2.2] Output left to be processed, not cleaning job yet.

```

Python output

```
Neomake [+53.3]: [8] Running makers: python (auto), frosted (auto), pyflakes (auto), pep8 (auto), pydocstyle (auto), pylint (auto)
Neomake [     ]: "python" "C:\Users\MY_USERNAME\vimfiles\pack\all\start\neomake\autoload\neomake\makers\ft\python\compile.py" "c:\Users\MY_USERNAME\neomake_py_lint.py"
Neomake [     ]: C:\WINDOWS\system32\cmd.exe /c """python"" ""C:\Users\MY_USERNAM\vimfiles\pack\all\start\neomake\autoload\neomake\makers\ft\python\compile.py"" ""c:\Users\MY_USERNAM\neomake_py_lint.py"""
Neomake [     ]: [8] Starting async job: 'C:\WINDOWS\system32\cmd.exe /c """python"" ""C:\Users\MY_USERNAM\vimfiles\pack\all\start\neomake\autoload\neomake\makers\ft\python\compile.py"" ""c:\Users\MY_USERNAM\neomake_py_lint.py"""'
Neomake [     ]: [8.7] Vim job: {'status': 'run', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel 7 open, 'process': 17500}
Neomake [     ]: [8.7] Vim channel: {'status': 'open', 'id': 7, 'in_io': 'pipe', 'err_mode': 'RAW', 'in_status': 'open', 'out_io': 'pipe', 'err_io': 'pipe', 'err_status': 'open', 'out_timeout': 2000, 'out_status': 'open', 'out_mode': 'RAW', 'in_mode': 'RAW', 'in_timeout': 2000, 'err_timeout': 2000}
```

With the patch I get no response from any Python linters but it does look like they're running:

```
19:37:46 [V      ] [2] Starting async job: 'C:\Windows\system32\cmd.exe /c """pylama"" ""--format"" ""parsable"" ""C:\Users\kyle\projects\test\test.py"""'
19:37:46 [D      ] [2.5] Vim job: {'status': 'run', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel 5 open, 'process': 6156}
19:37:46 [D      ] [2.5] Vim channel: {'status': 'open', 'id': 5, 'in_io': 'pipe', 'err_mode': 'RAW', 'in_status': 'open', 'out_io': 'pipe', 'err_io': 'pipe', 'err_status': 'open', 'out_timeout': 2000, 'out_status': 'open', 'out_mode': 'RAW', 'in_mode': 'RAW', 'in_timeout': 2000, 'err_timeout': 2000}
19:37:47 [D +1.10] [2.5] exit: pylama: 0
19:37:47 [D      ] [2.5] pylama: completed with exit code 0.
```

With Node linters an error message is shown in the location list:

```
|| module.js:471
|| throw err;
|| ^
|| 
|| Error: Cannot find module 'C:\Users\kyle\projects\test\node_modules\eslint\bin\eslint.js'
|| at Function.Module._resolveFilename (module.js:469:15)
|| at Function.Module._load (module.js:417:25)
|| at Module.runMain (module.js:604:10)
|| at run (bootstrap_node.js:394:7)
|| at startup (bootstrap_node.js:149:9)
|| at bootstrap_node.js:509:3
```

Notice that it's looking for a local eslint under the current pwd instead of using the one from PATH.

And from the neomake log:

```
19:38:59 [D +0.02] "eslint" "-f" "compact" "C:\Users\kyle\projects\test\test.js"
19:38:59 [D      ] C:\Windows\system32\cmd.exe /c """eslint"" ""-f"" ""compact"" ""C:\Users\kyle\projects\test\test.js"""
19:38:59 [V      ] [3] Starting async job: 'C:\Windows\system32\cmd.exe /c """eslint"" ""-f"" ""compact"" ""C:\Users\kyle\projects\test\test.js"""'
19:38:59 [D      ] [3.9] Vim job: {'status': 'run', 'stoponexit': 'term', 'exitval': 0, 'exit_cb': 0, 'channel': channel 9 open, 'process': 9656}
19:38:59 [D      ] [3.9] Vim channel: {'status': 'open', 'id': 9, 'in_io': 'pipe', 'err_mode': 'RAW', 'in_status': 'open', 'out_io': 'pipe', 'err_io': 'pipe', 'err_status': 'open', 'out_timeout': 2000, 'out_status': 'open', 'out_mode': 'RAW', 'in_mode': 'RAW', 'in_timeout': 2000, 'err_timeout': 2000}
19:38:59 [D +0.19] [3.9] stderr: eslint: ['module.js:471', '    throw err;', '    ^', '', 'Error: Cannot find module ''C:\Users\kyle\projects\test\node_modules\eslint\bin\eslint.js''', '    at Function.Module._resolveFilename (module.js:469:15)', '    at Function.Module._load (module.js:417:25)', '    at Module.runMain (module.js:604:10)', '    at run (bootstrap_node.js:394:7)', '    at startup (bootstrap_node.js:149:9)', '    at bootstrap_node.js:509:3', '']
```

--- a/autoload/neomake.vim
+++ b/autoload/neomake.vim
@@ -1540,9 +1540,8 @@ function! neomake#DisplayInfo() abort
         unlet! V  " Fix variable type mismatch with Vim 7.3.
     endfor
     echo "\n"
-    echo 'shell:' &shell
-    echo 'shellcmdflag:' &shellcmdflag
     echo 'Windows: '.neomake#utils#IsRunningWindows()
+    echo '[shell, shellcmdflag, shellslash]:' [&shell, &shellcmdflag, &shellslash]
     echo '```'
     if &verbose
         echo "\n"

--- a/autoload/neomake.vim
+++ b/autoload/neomake.vim
@@ -348,31 +348,35 @@ function! s:maker_base._get_argv(...) abort dict
         endif
     endif
 
-    if neomake#has_async_support()
+    if has('nvim')
         if args_is_list
             let argv = [exe] + args
         else
             let argv = exe . (len(args) ? ' ' . args : '')
         endif
-        if !has('nvim')
-            if !args_is_list
-                " Have to use a shell to handle argv properly (Vim splits it
-                " at spaces).
-                let argv = [&shell, &shellcmdflag, argv]
+    elseif neomake#has_async_support()
+        " Vim jobs, need special treatment on Windows..
+        if neomake#utils#IsRunningWindows()
+            " Windows needs a subshell to handle PATH/%PATHEXT% etc.
+            if args_is_list
+                let argv = join(map(copy([exe] + args), 'neomake#utils#shellescape(v:val)'))
+            else
+                let argv = exe.' '.args
             endif
+            let argv = &shell.' '.&shellcmdflag.' '.argv
+
+        elseif !args_is_list
+            " Use a shell to handle argv properly (Vim splits at spaces).
+            let argv = [&shell, &shellcmdflag, exe.' '.args]
+        else
+            let argv = [exe] + args
         endif
     else
-        let argv = exe
-        if len(args)
-            if args_is_list
-                if neomake#utils#IsRunningWindows()
-                    let argv .= ' '.join(args)
-                else
-                    let argv .= ' '.join(map(args, 'shellescape(v:val)'))
-                endif
-            else
-                let argv .= ' '.args
-            endif
+        " Vim-async, via system().
+        if args_is_list
+            let argv = join(map(copy([exe] + args), 'neomake#utils#shellescape(v:val)'))
+        else
+            let argv = exe.' '.args
         endif
     endif
     return argv

--- a/autoload/neomake/utils.vim
+++ b/autoload/neomake/utils.vim
@@ -468,3 +468,18 @@ endfunction
 function! neomake#utils#JSONdecode(json) abort
     return neomake#compat#json_decode(a:json)
 endfunction
+
+" Smarter shellescape, via vim-fugitive.
+function! s:gsub(str,pat,rep) abort
+  return substitute(a:str,'\v\C'.a:pat,a:rep,'g')
+endfunction
+
+function! neomake#utils#shellescape(arg) abort
+  if a:arg =~# '^[A-Za-z0-9_/.-]\+$'
+    return a:arg
+  elseif &shell =~? 'cmd' || exists('+shellslash') && !&shellslash
+    return '"'.s:gsub(s:gsub(a:arg, '"', '""'), '\%', '"%"').'"'
+  else
+    return shellescape(a:arg)
+  endif
+endfunction

--- a/tests/args.vader
+++ b/tests/args.vader
@@ -1,6 +1,6 @@
 Include: include/setup.vader
 
-Execute (maker.args as string gets not escaped):
+Execute (maker.args as list gets escaped):
   let maker = {
         \ 'exe': '/usr/bin/printf',
         \ 'args': ['%s\n', '1', '2', '3a 3b'],
@@ -17,7 +17,18 @@ Execute (maker.args as string gets not escaped):
   AssertEqual map(copy(getloclist(0)), 'v:val.text'),
     \ ['1', '2', '3a 3b', fname_abs]
 
-  let maker.args = "'%s\\n' 1 2 '3a 3b' 4"
+Execute (maker.args as string gets not escaped):
+  let maker = {
+        \ 'exe': '/usr/bin/printf',
+        \ 'args': "'%s\\n' 1 2 '3a 3b' 4",
+        \ 'errorformat': '%f: %m',
+        \ 'mapexpr': 'expand("%:p") . ": " . v:val',
+        \ }
+
+  let fname = 'tests/fixtures/a filename with spaces'
+  e! tests/fixtures/a\ filename\ with\ spaces
+  let fname_abs = expand('%:p')
+
   call neomake#Make(1, [maker])
   NeomakeTestsWaitForFinishedJobs
   AssertEqual map(copy(getloclist(0)), 'v:val.text'),

--- a/tests/neomake.vader
+++ b/tests/neomake.vader
@@ -300,5 +300,5 @@ Execute (Neomake picks up custom maker correctly):
     AssertNeomakeMessage "Starting async job: ['echo', '".fname."', '--foo', 'bar']"
     NeomakeTestsWaitForFinishedJobs
   else
-    AssertNeomakeMessage "Starting: echo '".fname."' '--foo' 'bar'"
+    AssertNeomakeMessage "Starting: echo ".fname." --foo bar"
   endif

--- a/tests/utils.vader
+++ b/tests/utils.vader
@@ -254,7 +254,7 @@ Execute (neomake#utils#MakerFromCommand splits shell/shellcmdflag):
   if neomake#has_async_support()
     let expected_argv = ['/bin/bash', '-o', 'pipefail', '-c', 'echo 1']
   else
-    let expected_argv = "/bin/bash '-o' 'pipefail' '-c' 'echo 1'"
+    let expected_argv = "/bin/bash -o pipefail -c 'echo 1'"
   endif
   let bound_maker = neomake#GetMaker(maker.fn(jobinfo))
   AssertEqual bound_maker._get_argv(), expected_argv
@@ -265,7 +265,7 @@ Execute (neomake#utils#MakerFromCommand splits shell/shellcmdflag):
   if neomake#has_async_support()
     let expected_argv = ['/bin/bash', '-o', 'pipefail', '-c', 'echo 2']
   else
-    let expected_argv = "/bin/bash '-o' 'pipefail' '-c' 'echo 2'"
+    let expected_argv = "/bin/bash -o pipefail -c 'echo 2'"
   endif
   let bound_maker = neomake#GetMaker(maker.fn(jobinfo))
   AssertEqual bound_maker._get_argv(), expected_argv

--- a/tests/verbosity.vader
+++ b/tests/verbosity.vader
@@ -71,7 +71,7 @@ Execute (neomake#Make uses &verbose):
     \ [[2, "Starting async job: ['echo', '1']", {'make_id': make_id}]]
   else
     AssertEqual g:neomake_test_messages,
-    \ [[2, "Starting: echo '1'", {'make_id': make_id}]]
+    \ [[2, "Starting: echo 1", {'make_id': make_id}]]
   endif
   AssertEqual len(g:neomake_test_messages), 1
 

